In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))


/kaggle/input/creditcardfraud/creditcard.csv


### Importing the libraries

In [35]:
data = pd.read_csv("/kaggle/input/creditcardfraud/creditcard.csv")
print(data)

            Time         V1         V2        V3        V4        V5  \
0       0.0      -1.359807  -0.072781   2.536347  1.378155 -0.338321   
1       0.0       1.191857   0.266151   0.166480  0.448154  0.060018   
2       1.0      -1.358354  -1.340163   1.773209  0.379780 -0.503198   
3       1.0      -0.966272  -0.185226   1.792993 -0.863291 -0.010309   
4       2.0      -1.158233   0.877737   1.548718  0.403034 -0.407193   
...     ...            ...        ...        ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0 -0.732789  -0.055080   2.035030 -0.738589  0.868229   
284804  172788.0  1.919565  -0.301254  -3.249640 -0.557828  2.630515   
284805  172788.0 -0.240440   0.530483   0.702510  0.689799 -0.377961   
284806  172792.0 -0.533413  -0.189733   0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.01830

In [36]:
pd.set_option('display.max_colwidth', -1)
data.head()

/tmp/ipykernel_32/2443526817.py:1: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option('display.max_colwidth', -1)


Time        V1        V2        V3        V4        V5        V6        V7  \
0  0.0  -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  0.0   1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2  1.0  -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3  1.0  -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4  2.0  -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62  0      
1  0.125895 -0.008983  0.014724  2.69    0      
2 -0.139097 -0.055353 -0.059752  378.66  0      
3 -0.221929  0.062723  0.061458  123.50  0      
4  0.502292  0.219422  0.215153  69.99   0      

[5 rows x 31 columns]

In [5]:
data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

### Getting some information about the dataset

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

### Checking the number of missing values in a dataset

In [7]:
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

### Checking the distribution of legit and fraudulent transactions

In [8]:
data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

### This dataset is highly unbalanced

### Separating the legitimate and Fraudulent transactions

In [9]:
legit = data[data.Class == 0]
fraud = data[data.Class == 1]
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


### Performing statistical tests on the data

In [10]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [11]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

### Comparing the values for both legitimate and fraudulent transactions

In [12]:
data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

### Dealing with the unbalanced data using under-sampling. Under sampling is building a sample dataset from the original dataset containing similar distribution of normal transactions and fraudulent transactions

### Number of fraudulent transactions = 492

In [13]:
legit_sample = legit.sample(n=492)

### Concatenating two dataframes, legit_sample and fraud. So now we will have equal amount of transactions in the new dataset

In [14]:
new_data = pd.concat([legit_sample, fraud],axis = 0)

In [15]:
new_data.head()

Time        V1        V2        V3        V4        V5        V6  \
257297  158093.0 -2.308209  0.516493 -0.465690 -0.361627  1.203851 -1.691980   
20072    30765.0  0.034085 -0.800402  1.036834  0.285444 -1.535104  1.602433   
14957    26265.0  1.128241  0.266870  0.558538  1.467770 -0.227972 -0.348162   
133225   80298.0  0.644878 -1.654490 -0.300843 -0.327368 -1.013153 -0.151153   
139297   83093.0  1.197730 -0.882558  0.943420 -0.758121 -1.288385  0.083174   

              V7        V8        V9  ...       V21       V22       V23  \
257297  0.831126 -0.220078  0.133948  ...  0.144765  0.846620 -0.574272   
20072  -2.395547 -2.241563 -0.120806  ... -0.969688  1.207318 -0.424330   
14957   0.154573 -0.097250  0.114547  ... -0.064168  0.083917 -0.040811   
133225  0.055662 -0.098589 -1.092611  ... -0.159181 -0.648810 -0.295442   
139297 -1.120926  0.189634 -0.714648  ...  0.492925  1.285840 -0.131134   

             V24       V25       V26       V27       V28  Amount  Class  
257297 -0.022222  0.384078 -0.160721 -0.221768  0.236070   21.71      0  
20072  -0.707002  0.619289  0.396587  0.167415  0.270615  150.23      0  
14957   0.436371  0.650857 -0.336605  0.045641  0.022667   15.68      0  
133225  0.080897  0.200178  1.067477 -0.122475  0.044986  356.07      0  
139297  0.072238  0.317713 -0.042817  0.043462  0.021086   59.00      0  

[5 rows x 31 columns]

In [16]:
new_data.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [17]:
new_data['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [18]:
new_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94416.997967  0.056250 -0.118836  0.108316  0.013115  0.025959   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.034574 -0.111420  0.026138  0.024505  ... -0.010282  0.011612   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.043059  0.004689  0.017149 -0.011244  0.027596 -0.016281  0.016721   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       81.932053  
1      122.211321  

[2 rows x 30 columns]

### Splitting the dataset into independent and independent variables

In [19]:
X = new_data.iloc[:, :-1].values
Y = new_data.iloc[:, -1].values
print(X)
print(Y)

[[ 1.58093000e+05 -2.30820931e+00  5.16493102e-01 ... -2.21768433e-01
   2.36069710e-01  2.17100000e+01]
 [ 3.07650000e+04  3.40853473e-02 -8.00402307e-01 ...  1.67415298e-01
   2.70614935e-01  1.50230000e+02]
 [ 2.62650000e+04  1.12824141e+00  2.66869743e-01 ...  4.56412000e-02
   2.26671138e-02  1.56800000e+01]
 ...
 [ 1.69351000e+05 -6.76142671e-01  1.12636606e+00 ...  3.85107449e-01
   1.94361479e-01  7.78900000e+01]
 [ 1.69966000e+05 -3.11383161e+00  5.85864172e-01 ...  8.84875540e-01
  -2.53700319e-01  2.45000000e+02]
 [ 1.70348000e+05  1.99197610e+00  1.58475887e-01 ...  2.98758224e-03
  -1.53088128e-02  4.25300000e+01]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

### Splitting into Training Data and Test Data

In [20]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, stratify = Y, random_state = 2)

### Using Logistic Regression

In [21]:
model = LogisticRegression()
model.fit(X_train, Y_train)

LogisticRegression()

### Checking the accuracy of the model on training data

In [26]:
X_train_pred = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_pred, Y_train)
print(f'The accuracy on training data is {training_data_accuracy * 100}%')

The accuracy on training data is 94.28208386277002%


### Checking the accuracy of the model on test data

In [27]:
X_test_pred = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_pred, Y_test)
print(f'The accuracy on test data is {test_data_accuracy * 100}%')

The accuracy on test data is 94.41624365482234%


### This model has a good accuracy of 94% on both training and test data, and can be used to detect legit and fradulent credit card transactions